include/albatross/src/indexing/traits.hpp

```diff
@@ -0,0 +1,195 @@
+};
+
+template <typename GrouperFunction, typename ValueType>
+struct group_key_is_valid {
+  static constexpr bool value = is_valid_map_key<
+      typename grouper_return_type<GrouperFunction, ValueType>::type>::value;
+};
+
+template <typename GrouperFunction, typename ValueType>
+struct is_valid_grouper {
+
+  static constexpr bool value =
+      can_be_called_with_const_ref<GrouperFunction, ValueType>::value &&
+      group_key_is_valid<GrouperFunction, ValueType>::value;
+};
+
+/*
+ * An ApplyFunction is a function which takes a key value pair and
+ * returns a new type, they key type needs to remain unchanged but
+ * the value can be modified.
+ */
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct key_value_apply_return_type
+    : public return_type_when_called_with<
+          ApplyFunction, typename const_ref<KeyType>::type, ArgType> {};
+
+template <typename ApplyFunction, typename ArgType>
+struct value_only_apply_return_type
+    : public return_type_when_called_with<ApplyFunction, ArgType> {};
+
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct is_valid_key_value_apply_function
+    : public can_be_called_with_const_ref<
+          ApplyFunction, typename const_ref<KeyType>::type, ArgType> {};
+
+template <typename ApplyFunction, typename ArgType>
+struct is_valid_value_only_apply_function
+    : public can_be_called_with_const_ref<ApplyFunction, ArgType> {};
+
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct is_valid_index_apply_function
+    : public can_be_called_with_const_ref<
+          ApplyFunction, typename const_ref<KeyType>::type,
+          typename const_ref<GroupIndices>::type> {};
+
+template <typename FilterFunction, typename... Args>
+struct returns_bool_when_called_with {
+  static constexpr bool value = std::is_same<
+      typename return_type_when_called_with<FilterFunction, Args...>::type,
+      bool>::value;
+};
+
+template <typename FilterFunction, typename KeyType, typename ArgType>
+struct is_valid_key_value_filter_function {
+  static constexpr bool value =
+      can_be_called_with_const_ref<FilterFunction, KeyType, ArgType>::value &&
+      returns_bool_when_called_with<FilterFunction, KeyType, ArgType>::value;
+};
+
+template <typename FilterFunction, typename ArgType>
+struct is_valid_value_only_filter_function {
+  static constexpr bool value =
+      can_be_called_with_const_ref<FilterFunction, ArgType>::value &&
+      returns_bool_when_called_with<FilterFunction, ArgType>::value;
+};
+
+/*
+ * The following traits are required in order to allow inspection of
+ * the only partially defined Derived types inside of GroupByBase.
+ *
+ * To get GroupByBase to work with other types you need to add a new
+ * trait struct for the type.
+ *
+ * GrouperFunction is a callable type which should take
+ */
+template <typename T> struct traits {};
```

Comment:
Would it make sense to call this something like `group_by_traits`?

Reply:
yeah, good call.